gobblin-api/src/main/java/gobblin/configuration/State.java

```diff
@@ -297,4 +296,9 @@ public void write(DataOutput out) throws IOException {
   public String toString() {
     return properties.toString();
   }
+
+  @Override
+  public int hashCode() {
+    return id == null ? this.properties.hashCode() : this.properties.hashCode() + this.id.hashCode();
```

Comment:
Please have a look at http://stackoverflow.com/questions/113511/best-implementation-for-hashcode-method for a reasonably good implementation of hashCode().